src/runtime/nitro/plugins/03-subresourceIntegrity.ts

```diff
@@ -2,6 +2,8 @@ import { useStorage, defineNitroPlugin, getRouteRules } from '#imports'
 import { isPrerendering } from '../utils'
 import { type CheerioAPI } from 'cheerio'
 
+const SCRIPT_RE = /<script((?=[^>]+src="[\w:.-\\/]+")(?![^>]+integrity="[\w-]+")[^>]+)\/>/g
```

Comment:
Progress report on the script regex:
```ts
const SCRIPT_RE = /<script((?=[^>]+src="[\w:.-\\/]+")(?![^>]+integrity="[\w-]+")[^>]+)(?:\/>|><\/script>)/g

const a =[
  `<script src="a.k"/><script src="b"></script>`,
  `<div>`,
  `<script>c</script>`,
  `<script src="a" integrity="a"></script> <script src="b"/>`
]
console.log(a.map(r=>r.replace(SCRIPT_RE,(match, mb)=>{return `<script integrity="a"${mb}></script>`})))
```
![image](https://github.com/Baroshem/nuxt-security/assets/92037085/b313657c-3cf4-49da-94ae-bd83a16b8ffb)


It works for both self-closing and non-self-closing script tags and automatically converts them to non-self-closing.

Reply:
Very nice !
Can you test also for
- integrity attribute before src attribute
- random attribute (e.g. `data-xxx`) before and after src attribute
- no src attribute with self-closing syntax
As well as edge cases such as :
- `<script>console.log('I want to mess up with javascript by printing </script>')</script>`
- `<p> I like to mess up with pre tags such as <pre> a code example: <script></script> </pre> </p>`
- `<script><!-- I like to mess up with HTML comments such as </script> --></script>

Reply:
Will do and post pictures soon!

Reply:
![image](https://github.com/Baroshem/nuxt-security/assets/92037085/251d3605-54e8-4996-b4db-b5735e1e0621)


Tested with this sample:
```ts
const SCRIPT_RE = /<script((?=[^>]+src="([\w:.-\\/]+)")(?![^>]+integrity="[\w-]+")[^>]+)(?:\/>|><\/script>)/g
const sriHashes = {
  "a.k":"a",
  "b":"c",
  "script.js":"newIntegrity"
}
const a = [
  `<script src="a.k"/><script src="b"></script>`,
  `<div>`,
  `<script>c</script>`,
  `<script src="a" integrity="a"></script> <script src="b"/>`,
  `<script integrity="integrity" src="script.js"></script>`, //integrity before src
  `<script data-xxx src="script.js"></script>`, //random attr before
  `<script src="script.js" data-yyy></script>`, //random attr after
  `<script data-xxx src="script.js" data-yyy></script>`, //random attr before and after
  `<script src="script1.js"></script><script src="script2.js"></script>`, //multiple scripts in the same line
  `<script>console.log('I like to mess up with javascript by printing </script>')</script>`, //first edge case - print
  `<p> I like to mess up with pre tags such as <pre> a code example: <script></script> </pre> </p>`, // second edge case - <pre>
  `<p> I like to mess up with pre tags such as <pre> a code example: <script> that is non-closing by mistake </pre> </p>`, // second edge case - <pre> (evil version)
  `<script><!-- I like to mess up with HTML comments such as </script> --></script>`, // third edge case - comment inside script
  `<script><!-- I like to mess up with HTML comments such as <script> --></script>` // third edge case - comment inside script (variant)
];
console.log(a.map(r=>r.replace(SCRIPT_RE,(match, rest, src: Partial<keyof typeof sriHashes>)=>{
  const hash = sriHashes[src]
  if (hash) {
    return `<script integrity="${hash}"${rest}></script>`
  }
  return match
})))
```
        
Please let me know if anything got incorrect results!

Reply:
Amazing, thanks a lot !!

Reply:
Glad to help! Of course I will also add those tests to e2e tests in the repo

Reply:
The real issue would be the inline scripts, I would need to think about how to make the regex robust enough.

Reply:
As per the cache, what do you think about using a `cache` key in the event.context? A use case for that would be the fact that for external scripts in 04, for example, we won't need to re-test it with a regex. We could save a WeakMap when we go through 03, where each script string would have the integrity as the value. That way we could simply get it from the map in stead of making even more regex operations. Is there anything that could hinder it as far as you know?

Reply:
Also, one more thing. Would it be too much of an issue if "valid" external scripts would get integrity? I just tested with `console.log(<script src="a.k"></script>)` and it returned the the script with the integrity, but I don't really think that would cause any issues, would it?

Reply:
Sounds good to me

Reply:
Yea, unfortunately you can't win them all with regex.. But at least the edge cases and potential issues are covered, so that's something

Reply:
> Also, one more thing. Would it be too much of an issue if "valid" external scripts would get integrity? I just tested with `console.log(<script src="a.k"></script>)` and it returned the the script with the integrity, but I don't really think that would cause any trouble, would it?

I am not too sure to be honest
- On the one hand, in theory this is a problem that could lead to many issues. We are modifying user code by doing XSS ourselves here, which looks really bad for a security module
- On the other hand, I don't even know how browsers are supposed to interpret this... 

This might be the reason why the author of https://github.com/Baroshem/nuxt-security/blob/v1.0.0-rc.1/src/runtime/nitro/plugins/99-cspNonce.ts was doing some lookbehind search around quotes

Reply:
> Yea, unfortunately you can't win them all with regex.. But at least the edge cases and potential issues are covered, so that's something

Sorry, I was commenting on your idea of a WeakMap cache
Our comments on integrity injection crossed at the same time

Reply:
I see, all good!

Reply:
> This might be the reason why the author of https://github.com/Baroshem/nuxt-security/blob/v1.0.0-rc.1/src/runtime/nitro/plugins/99-cspNonce.ts was doing some lookbehind search around quotes

Yea, but at the same time it could be something like:
<div> hi'<script>console.log('hi')</script></div>

Reply:
It's not that straight-forward unfortunately